### PR TITLE
Nonparametric two graph testing and rho-er + rho-sbm

### PR DESCRIPTION
**Pull request by bvarjavand**
#### Reference Issues/PRs

Delivering nonparametric testing and rho-sbm.

#### What does this implement/fix? Explain your changes.
Implements methods, including tutorials and tests.

### CODE REVIEW

File: graspy/inference/nonpar.py [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+        A2 = import_graph(A2)
+        if not is_symmetric(A1) or not is_symmetric(A2):
+            raise NotImplementedError()  # TODO asymmetric case
+
+        if self.n_components is None:
+            num_dims1 = select_dimension(A1)[0][-1]
+            num_dims2 = select_dimension(A2)[0][-1]
+            self.n_components = max(num_dims1, num_dims2)
+
+        # monte = [] TODO
```
Comment (bdpedigo) [inline graspy/inference/nonpar.py]: what is the n_monte?

File: graspy/inference/nonpar.py [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+
+    def _dist_matrix(self, z):
+        dists = pdist(z, 'euclidean')
+        zlen = len(z)
+        ind = np.triu_indices(zlen, k=1)
+        k = np.zeros((zlen, zlen))
+        k[ind] = dists
+        return symmetrize(k)
+
+    def _embed(self, A1, A2):
```
Comment (bdpedigo) [inline graspy/inference/nonpar.py]: If this is supposed to be the exact same as semipar we should refactor to the base class. Note that I don't think LSE makes sense for semipar so I took it out, have you tested it for nonpar?

File: graspy/simulations/simulations.py [hunk 3]
```diff
@@ -428,15 +428,15 @@ def rdpg(X,
     Returns
     -------
     P: np.ndarray (X.shape[0], X.shape[0])
-        A matrix representing the probabilities of connections between 
+        A matrix representing the probabilities of connections between
         vertices in a random graph based on their latent positions
 
     References
     ----------
     .. [1] Sussman, D.L., Tang, M., Fishkind, D.E., Priebe, C.E.  "A
```
Comment (bdpedigo) [inline graspy/simulations/simulations.py]: update reference

File: graspy/inference/nonpar.py (hunk 2 above)
Comment (bvarjavand) [reply graspy/inference/nonpar.py]: You're absolutely right

File: graspy/inference/nonpar.py (hunk 1 above)
Comment (bvarjavand) [reply graspy/inference/nonpar.py]: vestigial code - removing

File: graspy/inference/nonpar.py [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+        for _ in range(self.n_bootstraps):
+            np.random.shuffle(z)
+            u = self._u_from_z(z, len(x))
+            boots.append(u)
+        return boots
+
+    def _u_from_z(self, z, n):
+        k = self._dist_matrix(z)
+        gk = np.exp(-k / np.median(k))
+        u = np.mean(gk[:n, :n]) + np.mean(gk[n:, n:]) - 2 * np.mean(gk[:n, n:])
```
Comment (j1c) [inline graspy/inference/nonpar.py]: According to eq 2.2 in the paper, this is the biased U

File: graspy/inference/nonpar.py [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+        boots = []
+        for _ in range(self.n_bootstraps):
+            np.random.shuffle(z)
+            u = self._u_from_z(z, len(x))
+            boots.append(u)
+        return boots
+
+    def _u_from_z(self, z, n):
+        k = self._dist_matrix(z)
+        gk = np.exp(-k / np.median(k))
```
Comment (j1c) [inline graspy/inference/nonpar.py]: Purpose of the exponential? Something Avanti said?

**Pull request closed without merging**